[5.0.0-rc2] Backport Fix chrome/selenium tests (#25330)

JunTaoLuo opened the pull request:
Back port of https://github.com/dotnet/aspnetcore/pull/25330 to rc2. Since it's only testing, I'm going to mark this as tell mode.

Commit:
Fix chrome/selenium tests (#25330)

* Revert "Disable failing/hanging tests due to Chrome/Selenium issue (#25323)"

This reverts commit 332f1504128a63431cb67830703922cdeefee525.

* Update Selenium to latest

* Update API

* Try specifying a version

* Update Selenium to 4.0.0-beta5

* Disable browser log tests

* Fix components e2e tests and disable blazor standalone template test

dougbu commented:
Approved but it appears different tests are failing on this branch

JunTaoLuo commented:
I see one rc2 test that's affected by browser logs which I can disable. There's one blazor pwa template I'm not sure about and the rest look like setting cultures is broken. Not sure why those would be broken.

Commit:
Disable tests using browser log

JunTaoLuo commented:
Taking a quick look, it does seem like there are some test issues, at least for the culture-related ones.

For example:
```
Assert.Equal() Failure
                      ↓ (pos 12)
Expected: 02.09.2020 0:00:00
Actual:   02.09.2020 00:00:00
                      ↑ (pos 12)

Assert.Equal() Failure
              ↓ (pos 4)
Expected: 2020-09-02 오전 12:00:00
Actual:   2020. 9. 2. 오전 12:00:00
              ↑ (pos 4)
```

SteveSandersonMS commented on src/Components/test/testassets/TestServer/ServerStartup.cs:
```diff
@@ -19,7 +19,7 @@ public ServerStartup(IConfiguration configuration)
         {
             services.AddMvc();
-            services.AddServerSideBlazor(options => options.DetailedErrors = true);
+            services.AddServerSideBlazor();
```
Out of interest, why is this change required?

JunTaoLuo replied:
This option controls how some exceptions are logged: https://github.com/dotnet/aspnetcore/blob/ec2c7b478769c68870ec860592be27cc8920519b/src/Components/Server/src/Circuits/RemoteJSRuntime.cs#L43-L56
and it turns out some tests check for this behaviour: https://github.com/dotnet/aspnetcore/blob/9a1810c1dbe432fc7bc7e8bc68fa22ab787c0452/src/Components/test/E2ETest/ServerExecutionTests/ServerInteropTestDefaultExceptionsBehavior.cs#L66-L68

given there were several tests that seemed to be dependent on this options, I just removed it. It's possible that you might consider changing the log messages in the tests instead.

SteveSandersonMS replied:
This option affects the exception messages that are returned to the browser in JS interop calls. It doesn't only affect what gets logged to the browser console, it affects the exceptions that can be caught by JS code. The test code looks at the exceptions caught by JS code.

So I'm unclear on why this needs to be changed. I thought the only issue was with reading the browser console logs with the new version of Selenium, but the test in question (`DotNetExceptionDetailsAreNotLoggedByDefault`) wasn't doing that anyway.

> It's possible that you might consider changing the log messages in the tests instead.

Could you clarify what you're suggesting?

JunTaoLuo replied:
No this has nothing to do with the browser console logs. Here the issue is that the test validates the exception message which is controlled by `DetailedErrors`. During the time we turned off e2etests, this option was modified to be true and therefore affected the exception message. The tests weren't updated in response to setting `DetailedErrors` to true and we saw failures when I re-enabled the tests. I opted to set `DetailedError` back to what it was before but the suggestion is that we can modify the exception message we verify to match the scenario where `DetailedError` is true as an alternative fix.

SteveSandersonMS replied:
That makes sense! Thanks for explaining. All LGTM here then!

SteveSandersonMS reviewed (approved):
Changes look good to me. Thanks for pursuing this, @JunTaoLuo!

As for why certain culture-specific tests are failing, I'm unsure. @pranavkm, you recently did some work to set up culture-specific E2E tests. Do you have any insights into what's going on here?

SteveSandersonMS reviewed (commented):
Seeking clarification on comment above

Commit:
Disable template test

pranavkm committed:
Avoid using .NET formatted strings in tests

* Annotate BasicTestApp suggesting that it needs the all globalization data

* Culture specific formatting relies on the ICU data carried by the OS. This
causes issues in our tests if WebAssembly carries a different set than the OS. Instead
updating these tests to use hardcoded strings.

* Additionally fixing an issue where some projects in the solution were using tasks from
the .dotnet SDK rather than the local copy of the SDK. This was causing issues building locally.

pranavkm commented on eng/targets/CSharp.Common.props:
```diff
@@ -12,7 +12,12 @@
 
-  <ItemGroup Condition="'$(UsingMicrosoftNETSdkWeb)' == 'true' OR '$(UsingMicrosoftNETSdkBlazorWebAssembly)' == 'true' OR '$(RazorSdkCurrentVersionProps)' != ''">
+  <PropertyGroup>
+    <_ReferenceLocalRazorSDK
```
@SteveSandersonMS this should fix the issue with you were seeing. I ran in to when debugging the tests in VS and really had to fix this to make any progression.

pranavkm commented on src/Components/test/E2ETest/Tests/WebAssemblyICUShardingTest.cs:
```diff
@@ -39,7 +36,7 @@ public void LoadingApp_FrenchLanguage_Works()
 
             var dateDisplay = Browser.Exists(By.Id("dateTime"));
-            Assert.Equal(DisplayTime.ToString(culture), dateDisplay.Text);
+            Assert.Equal("02/09/2020 00:00:00", dateDisplay.Text);
```
The ICU data on the server depends on the OS. On my machine, it's up to date with WASM but not so on the build machines. Using hardcoded strings seems like the best way to proceed: https://github.com/dotnet/runtime/pull/42136

JunTaoLuo commented:
Thanks for the fix @pranavkm
